hazelcast/serialization/compact.py

```diff
@@ -0,0 +1,2240 @@
+import abc
+import collections
+import datetime
+import decimal
+import typing
+
+from hazelcast.errors import HazelcastError, HazelcastSerializationError, IllegalStateError
+from hazelcast.serialization import (
+    BYTE_SIZE_IN_BYTES,
+    CHAR_SIZE_IN_BYTES,
+    SHORT_SIZE_IN_BYTES,
+    INT_SIZE_IN_BYTES,
+    LONG_SIZE_IN_BYTES,
+    FLOAT_SIZE_IN_BYTES,
+    DOUBLE_SIZE_IN_BYTES,
+)
+from hazelcast.serialization.api import (
+    CompactSerializer,
+    ObjectDataOutput,
+    CompactWriter,
+    CompactReader,
+    FieldKind,
+    ObjectDataInput,
+)
+from hazelcast.serialization.input import _ObjectDataInput
+from hazelcast.serialization.output import _ObjectDataOutput
+from hazelcast.serialization.serialization_const import TYPE_COMPACT
+from hazelcast.serialization.serializer import BaseSerializer
+from hazelcast.serialization.util import IOUtil
+
+_BOOLEANS_PER_BYTE = 8
+
+
+class CompactStreamSerializer(BaseSerializer):
+    def __init__(self, compact_serializers: typing.Dict[typing.Type, CompactSerializer]):
+        self._type_to_serializer = compact_serializers
+        self._type_name_to_serializer: typing.Dict[str, CompactSerializer] = {
+            serializer.get_type_name(): serializer for serializer in compact_serializers.values()
+        }
+        self._type_to_schema: typing.Dict[typing.Type, Schema] = {}
+        self._id_to_schema: typing.Dict[int, Schema] = {}
+
+    def write(self, out: ObjectDataOutput, obj: typing.Any) -> None:
+        clazz = type(obj)
+        serializer = self._type_to_serializer[clazz]
+        schema = self._type_to_schema.get(clazz)
+        if not schema:
+            schema = CompactStreamSerializer._build_schema(serializer, obj)
+            # Check if we already fetched the schema from the cluster.
+            if schema.schema_id not in self._id_to_schema:
+                raise SchemaNotReplicatedError(schema, clazz)
+
+            # No need to raise the not replicated error, as we just
+            # fetched it from the cluster. Let's register it to our
+            # local and continue
+            self.register_sent_schema(schema, clazz)
+
+        out.write_long(schema.schema_id)
+        writer = DefaultCompactWriter(self, out, schema)  # type: ignore
+        serializer.write(writer, obj)
+        writer.write_var_sized_field_positions()
+
+    def read(self, inp: ObjectDataInput) -> typing.Any:
+        schema_id = inp.read_long()
+        schema = self._id_to_schema.get(schema_id)
+        if not schema:
+            raise SchemaNotFoundError(schema_id)
+
+        serializer = self._type_name_to_serializer.get(schema.type_name)
+        if not serializer:
+            raise HazelcastSerializationError(
+                f"No serializer is registered for the type name '{schema.type_name}'"
```

Comment:
Maybe saying `No compact serializer is registered` is better for users to understand the data is in compact format.